petesong/Add a Python example for the test practice

PeteSong opened the pull request:
### **User description**
**Thanks for contributing to the Selenium site and documentation!**
**A PR well described will help maintainers to review and merge it quickly**

Before submitting your PR, please check our [contributing](https://selenium.dev/documentation/en/contributing/) guidelines.
Avoid large PRs, and help reviewers by making them as simple and short as possible.


<!--- Provide a general summary of your changes in the Title above -->

### Description
1. Correct the right link of Hugo Installation
2. Add a Python example for the test practices.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
For the test practices, there's no python example. 

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Change to the site 
- [ ] Code example added (and I also added the example to all translated languages)
- [ ] Improved translation
- [ ] Added new translation (and I also added a notice to each document missing translation)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [**contributing**](https://selenium.dev/documentation/en/contributing/) document.
- [x] I have used `hugo server` to render the site/docs locally, but failed. Since I only have minor changes in the existing pages,  and I am sure it works.
<!--- Provide a general summary of your changes in the Title above -->


___

### **PR Type**
Documentation, Tests


___

### **Description**
- Updated the Hugo installation link in the README to point directly to the official Hugo site, improving clarity and accessibility.
- Added a detailed Python example using pytest and Selenium to the test practices documentation. This example demonstrates best practices with the "ActionBot" and "LoadableComponent" patterns.
- The example includes multiple test cases for interacting with a Todo application, showcasing various functionalities such as adding, toggling, and deleting todos.



___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Update Hugo installation link and instructions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

README.md

<li>Updated the Hugo installation link to a more direct source.<br> <li> Provided clearer instructions for getting started with Hugo.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/seleniumhq.github.io/pull/1924/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>design_strategies.en.md</strong><dd><code>Add Python example for test practices with Selenium</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

website_and_docs/content/documentation/test_practices/design_strategies.en.md

<li>Added a comprehensive Python example using pytest and Selenium.<br> <li> Demonstrated best practices with "ActionBot" and "LoadableComponent".<br> <li> Included multiple test cases for a Todo application.<br>


</details>


  </td>
  <td><a href="https://github.com/SeleniumHQ/seleniumhq.github.io/pull/1924/files#diff-ef76a0a38d2ead69e3cef4401392425717c886b9e7b638ede34872bf953aec22">+280/-0</a>&nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

PeteSong committed:
Fix the Hugo installation link

PeteSong committed:
Add an example for the test practices

Use `python + pytest + selenium` to test the Todo page with the best test practices: "ActionBot" and "LoadableComponent"

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/SeleniumHQ/seleniumhq.github.io?pullRequest=1924) <br/>All committers have signed the CLA.

PeteSong committed:
Update design_strategies.en.md

As suggested,
1. Break the long and complex example down into smaller, more manageable sections or separate files for better readability and maintainability
2. Use a `with` context manager for WebDriver to ensure proper resource cleanup
3. Use specific exception handling instead of a bare except clause
4. Convert static method to instance method for consistency and flexibility
5. Add assertions to verify initial state in test methods
6. Remove the commented code in the `hover` method

PeteSong commented:
> ## PR Code Suggestions ✨
> Category	Suggestion                                                                                                                                   	Score
> **Best practice**	
> ✅ Use a context manager for WebDriver to ensure proper resource cleanup
> 9
> Use specific exception handling instead of a bare except clause
> 8
> ✅ Add assertions to verify initial state in test methods
> 7
> **Enhancement**	
> ✅ Convert static method to instance method for consistency and flexibility
> 5

Applied the suggestions.

PeteSong commented:
> ## PR Reviewer Guide 🔍
> ⏱️ **Estimated effort to review**: 4 🔵🔵🔵🔵⚪
> 🧪 **PR contains tests**
> 🔒 **No security concerns identified**
> ⚡ **Key issues to review**
> 
> [**Code Complexity**](https://github.com/SeleniumHQ/seleniumhq.github.io/pull/1924/files#diff-ef76a0a38d2ead69e3cef4401392425717c886b9e7b638ede34872bf953aec22R381-R657)The added Python example is quite long and complex. Consider breaking it down into smaller, more manageable sections or separate files for better readability and maintainability.
> 
> [**Error Handling**](https://github.com/SeleniumHQ/seleniumhq.github.io/pull/1924/files#diff-ef76a0a38d2ead69e3cef4401392425717c886b9e7b638ede34872bf953aec22R516-R523)The `is_loaded` method in the `TodoPage` class uses a bare except clause, which might catch and suppress unexpected errors. Consider using a more specific exception or at least logging the error.
> 
> [**Commented Code**](https://github.com/SeleniumHQ/seleniumhq.github.io/pull/1924/files#diff-ef76a0a38d2ead69e3cef4401392425717c886b9e7b638ede34872bf953aec22R430-R434)There is commented out code in the `hover` method of the `ActionBot` class. Consider removing it if it's not needed or add a comment explaining why it's kept.

Applied the suggestions.

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

shbenzer commented:
Thank you for the contribution @PeteSong! Before it can be properly reviewed a few things must be changed:
1. the example must be moved into a separate executable file like our other examples
2. the example must be properly formatted like other examples to pull from the example file
3. the example must be propagated to all translations
4. lines must be kept to 100 characters per the style guide https://www.selenium.dev/documentation/about/style/

PeteSong committed:
Create using_best_practice.py

example for using best practice

PeteSong committed:
Update design_strategies.zh-cn.md

PeteSong committed:
Update design_strategies.ja.md

PeteSong committed:
Update design_strategies.pt-br.md

PeteSong committed:
small update

PeteSong committed:
Merge branch 'petesong/add-practice-example' of https://github.com/PeteSong/seleniumhq.github.io into petesong/add-practice-example

PeteSong commented:
> Thank you for the contribution @PeteSong! Before it can be properly reviewed a few things must be changed:
> 
> 1. the example must be moved into a separate executable file like our other examples
> 2. the example must be properly formatted like other examples to pull from the example file
> 3. the example must be propagated to all translations
> 4. lines must be kept to 100 characters per the style guide https://www.selenium.dev/documentation/about/style/

DONE

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

shbenzer commented:
> > Thank you for the contribution @PeteSong! Before it can be properly reviewed a few things must be changed:
> > 
> > 1. the example must be moved into a separate executable file like our other examples
> > 2. the example must be properly formatted like other examples to pull from the example file
> > 3. the example must be propagated to all translations
> > 4. lines must be kept to 100 characters per the style guide https://www.selenium.dev/documentation/about/style/
> 
> DONE

Awesome, thank you! @harsha1502

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

A1exKH reviewed (changes requested):
@PeteSong please resolve conflicts for this PR.

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong commented:
> @PeteSong please resolve conflicts for this PR.

DONE

A1exKH reviewed (approved):
LGTM.

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong commented:
The merging is blocked, who could help me to merge it?

PeteSong commented:
@shbenzer, @A1exKH , 
Hope you are doing well. 

The merging is blocked for quite a time. What should I do or who could help me to merge it? 

Best regards,
Peter

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

A1exKH commented:
@PeteSong hi! 
You need a review and approval from people, who are contributors to this repository.

shbenzer commented:
Hello @PeteSong, thank you for your contributions!

I have added some review comments, let me know if you have any questions. The main point I'm trying to get across as you refactor the examples is that these are examples - they don't need to be fully fleshed out. 

Since you have provided a full seemingly working example in python - what we can _instead_ do is utilize the codeblock architecture:

```
{{< gh-codeblock path="examples/python/tests/selenium_manager/usage.py#L10-L12" >}}
{{< /tab >}}
{{< tab header="CSharp" >}}
{{< badge-code >}}
{{< /tab >}}
{{< tab header="Ruby" >}}
{{< badge-code >}}
{{< /tab >}}
{{% tab header="JavaScript" %}}
```

and therefore use all the code you've written throughout the page (I personally prefer this alternative) - and this way we can keep the java example, as well.

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
update "deisgn_strategies.en.md" to utilize `codeblock`

PeteSong committed:
update files to utilize `codeblock`

update the following files to utilize `codeblock`
1. "design_strategies.ja.md"
2. "design_strategies.pt-br.md"
3. "design_strategies.zh-cn.md"

PeteSong commented:
Hi @shbenzer ,

Thanks for your good comment. I have updated them per your suggestion. 

DONE!

Best regards,
Peter

> Hello @PeteSong, thank you for your contributions!
> 
> I have added some review comments, let me know if you have any questions. The main point I'm trying to get across as you refactor the examples is that these are examples - they don't need to be fully fleshed out.
> 
> Since you have provided a full seemingly working example in python - what we can _instead_ do is utilize the codeblock architecture:
> 
> ```
> {{< gh-codeblock path="examples/python/tests/selenium_manager/usage.py#L10-L12" >}}
> {{< /tab >}}
> {{< tab header="CSharp" >}}
> {{< badge-code >}}
> {{< /tab >}}
> {{< tab header="Ruby" >}}
> {{< badge-code >}}
> {{< /tab >}}
> {{% tab header="JavaScript" %}}
> ```
> 
> and therefore use all the code you've written throughout the page (I personally prefer this alternative) - and this way we can keep the java example, as well.

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong commented:
@shbenzer 
Hope you are doing well. Could you be kind to review the changes and approve the PR? Thank you very much

Best regards,
Peter

shbenzer commented:
Hey @PeteSong this looks good! Only thing I notice atm that needs to be changed is that you're missing the other languages in the codeblock elements. We use those so that readers can see where examples are missing and for ease of contributions.

PeteSong committed:
Add the other languages in the codeblock elements. So that readers can see where examples are missing and for ease of contributions.

PeteSong commented:
Hey @shbenzer, 

Good suggestion! DONE!

> Hey @PeteSong this looks good! Only thing I notice atm that needs to be changed is that you're missing the other languages in the codeblock elements. We use those so that readers can see where examples are missing and for ease of contributions.

shbenzer commented:
LGTM 

@harsha509 I don’t seem to have permissions to this repo, could you take a look?

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong commented:
Hi @harsha509, hope you are doing well!  Could I trouble you to take some time to review and merge it?  Thank you very much.

shbenzer commented:
Alright, now that I have permissions I pulled locally. There's a few more changes that need to be made and you'll be good to go.

1. lets put all examples in the codeblock elements
2. in the codeblock elements, make sure that each language without an example has a badgecode so we can help facilitate contributions

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong commented:
> Alright, now that I have permissions I pulled locally. There's a few more changes that need to be made and you'll be good to go.
> 
> 1. lets put all examples in the codeblock elements
> 2. in the codeblock elements, make sure that each language without an example has a badgecode so we can help facilitate contributions

Hi @shbenzer, for 1),  I already put all my examples in the codeblock elements. for 2), what is the badgecode? could you show me an example?

shbenzer commented:
> Hi @shbenzer, for 1), I already put all my examples in the codeblock elements. 

That's correct! However, now that we're using multiple languages in this document, it'd be best to put the other code examples you didn't alter into codeblocks so the page is consistently formatted.

> for 2), what is the badgecode? could you show me an example?

Gladly!

Badge codes have this structure in codeblocks:

```html
{{< tab header="JavaScript" >}}
{{< badge-code >}}
{{< /tab >}}
```

and they produce this when rendered:

<img width="639" alt="Screenshot 2025-05-04 at 12 40 27 PM" src="https://github.com/user-attachments/assets/8db52a0c-c80d-49db-b9c8-18ba203ccfa9" />

Clicking the link then takes you to the [creating examples section](https://www.selenium.dev/documentation/about/contributing/#creating-examples) of [the contribution guide](https://www.selenium.dev/documentation/about/contributing/)

PeteSong committed:
Merge branch 'trunk' into petesong/add-practice-example

PeteSong committed:
Merge branch 'petesong/add-practice-example' of https://github.com/PeteSong/seleniumhq.github.io into petesong/add-practice-example

PeteSong committed:
Added "badgecode"

PeteSong commented:
Hi @shbenzer,  hope you are doing well.
1. for now, only one example in Python. 
2. added the badgecode. 


Best regards,
Peter

shbenzer commented:
> Hi @shbenzer, hope you are doing well.
> 
> 1. for now, only one example in Python.
> 2. added the badgecode.
> 
> Best regards, Peter

Regarding 1, that's not what I meant, but it can be done in another PR.

LGTM

shbenzer committed:
Merge branch 'trunk' into petesong/add-practice-example

shbenzer commented:
Thank you @PeteSong !
